README.md

```diff
@@ -1,2 +1,151 @@
+
+## Create a Docker Machine
+
+Use [Docker Machine](https://docs.docker.com/machine/) to provision a new host, or to connect to an existing one.   
+
+This example assumes that a remote host already exists at IP address `10.0.0.10`, and that you can perform password-less login to this host from your local workstation or laptop using a private SSH key.
+
+To create a Docker machine that will install and configure the Docker daemon on an existing host (note that Docker Machine will upgrade the Docker daemon on the host if it is already installed):
+
+```
+# Use the generic driver to create a Docker machine that
+# controls the Docker daemon on an existing host
+export DRIVER_OPTS="--driver generic \
+  --generic-ip-address 10.0.0.10 \
+  --generic-ssh-key /path/to/private/ssh/key"
+
+# Create a machine named jupyterhub
+docker-machine create $DRIVER_OPTS jupyterhub
+
+# Activate the machine
+eval "$(docker-machine env jupyterhub)"
+```
+
+Docker Machine can also provision new hosts using one of it's [supported drivers](https://docs.docker.com/machine/drivers/).   When provisioning a host, Docker Machine will automatically generate TLS certificate and key files and use them to authenticate with the remote Docker daemon.
+
+## Create a Docker Network
+
+Create a Docker network for inter-container communication.  The benefits of using a Docker network are:
+
+* container isolation - only the containers on the network can access one another
+* name resolution - Docker daemon runs an embedded DNS server to provide automatic service discovery for containers connected to user-defined networks.  This allows us to access containers on the same network by name.
+
+Here we create a Docker network named `jupyterhub-network`.  Later, we will configure the JupyterHub and single-user Jupyter Notebook containers to run attached to this network.
+
+```
+docker network create jupyterhub-network
+```
+
+## Setup GitHub Authentication
+
+This deployment uses GitHub OAuth to authenticate users.  It requires that you create a [GitHub application](https://github.com/settings/applications/new). You will need to specify an OAuth callback URL in the following form:
+
+```
+https://<myhost.mydomain>/hub/oauth_callback
+```
+
+You must pass the secrets that GitHub provides for your application to JupyterHub at runtime.  You can do this by setting the `GITHUB_CLIENT_ID`, `GITHUB_CLIENT_SECRET`, and `OAUTH_CALLBACK_URL` environment variables when you run the JupyterHub container, or you can add them to the `.env` file in the root directory of this repository.  For example,
```

Comment:
We might want to add a note warning people that if they put these things in the `.env` file, that they must be very careful not to commit them. Or maybe tell them to put it in the `secrets` folder?